File: src/lib/api/message.ts
/**
 * 消息相关 API 封装
 */

import { apiCall } from './index';
import { listen } from '@tauri-apps/api/event';
import type { MessageRequest, MessageResponse, Message, UUID, MessageStreamEvent, ToolExecutionStatus, UserMessageSendRequest } from '../types';

/**
 * 发送消息
 */
export async function sendUserMessage(request: MessageRequest): Promise<MessageResponse> {
  // Tauri 命令期望参数名与函数参数名匹配
  const payload = {
    request: {
      chat_id: request.chatId,
      model_id: request.modelId,
      provider_id: request.providerId,
      messages: request.messages,
      attachments: request.attachments
    }
  };
  
  return await apiCall<any>('message_user_send', payload);
}

/**
 * 获取聊天下的消息列表
 */
export async function getMessages(
  chatId: UUID,
  limit?: number,
  offset?: number
): Promise<Message[]> {
  const list = await apiCall<any[]>('message_list', { chatId: chatId, limit, offset });
  console.log('getMessages >>> :', list);
  return list || [];
}

/**
 * 获取单条消息
 */
export async function getMessage(messageId: UUID): Promise<Message> {
  const r = await apiCall<any>('message_get', { messageId: messageId });
  return r;
}

/**
 * 更新消息
 */
export async function updateMessage(
  messageId: UUID,
  content: string
): Promise<Message> {
  const r = await apiCall<any>('message_update', { messageId: messageId, content });
  return r;
}

/**
 * 删除消息
 */
export async function deleteMessage(messageId: UUID): Promise<void> {
  return apiCall<void>('message_delete', { messageId: messageId });
}

/**
 * 流式重新生成助手消息 - 删除当前消息，根据本轮消息重新生成
 */
export async function regenerateAssistantMessageStream(messageId: UUID): Promise<void> {
  await apiCall<void>('message_assistant_regenerate_stream', { messageId: messageId });
}

/**
 * 流式重发用户消息 - 删除该消息之后的所有消息，然后重新发送（流式）
 * @param messageId 消息ID
 * @param content 可选的新消息内容，如果提供则更新消息内容后重新发送
 */
export async function resendUserMessageStream(messageId: UUID, content?: string): Promise<void> {
  await apiCall<void>('message_user_resend_stream', {
    messageId: messageId,
    content: content
  });
}

/**
 * 发送流式消息
 */
export async function sendUserMessageStream(request: UserMessageSendRequest): Promise<void> {
  // Tauri 命令期望参数名与函数参数名匹配
  const payload = {
    request: {
      chat_id: request.chatId,
      content: request.content,
      temp_user_message_id: request.tempUserMessageId,
      attachments: request.attachments
    }
  };

  await apiCall<void>('message_user_send_stream', payload);
}

/**
 * 监听流式消息事件
 */
export interface StreamEventHandlers {
  onStart?: (data: { streamId: string; messageId: string }) => void;
  onChunk?: (data: { streamId: string; content: string; reasoning?: string; toolCalls?: any[]; chunk: string; index: number }) => void;
  onEnd?: (data: { streamId: string; finalContent: string; finalReasoning?: string; chatId: string; modelId: string; providerId: string; toolCalls?: any[]; messageId?: string }) => void;
  onError?: (error: any) => void;
  onToolExecute?: (data: { messageId: string; toolCallIds: string[]; status: ToolExecutionStatus }) => void;
  onMessagesDelete?: (data: { chatId: string; messageIds: string[] }) => void;
  onUserMessageSaved?: (data: { tempMessageId: string; savedMessageId: string }) => void;
}

export async function listenToStreamEvents(handlers: StreamEventHandlers, eventPrefix: string = 'message_stream') {
  const listeners = [
    listen(`${eventPrefix}_start`, (event) => {
      handlers.onStart?.(event.payload as any);
    }),

    listen(`${eventPrefix}_chunk`, (event) => {
      handlers.onChunk?.(event.payload as any);
    }),

    listen(`${eventPrefix}_end`, (event) => {
      handlers.onEnd?.(event.payload as any);
    }),

    listen(`${eventPrefix}_error`, (event) => {
      handlers.onError?.(event.payload as any);
    })
  ];

  // 如果提供了 onToolExecute 处理器，添加工具执行事件监听
  if (handlers.onToolExecute) {
    console.log('[listenToStreamEvents] 注册 tool_execute 事件监听器');
    listeners.push(
      listen('tool_execute', (event) => {
        console.log('[tool_execute 事件] 收到事件:', event.payload);
        handlers.onToolExecute?.(event.payload as any);
      })
    );
  } else {
    console.warn('[listenToStreamEvents] 未提供 onToolExecute 处理器');
  }

  // 如果提供了 onMessagesDelete 处理器，添加消息删除事件监听
  if (handlers.onMessagesDelete) {
    console.log('[listenToStreamEvents] 注册 messages_deleted 事件监听器');
    listeners.push(
      listen('messages_deleted', (event) => {
        console.log('[messages_deleted 事件] 收到事件:', event.payload);
        handlers.onMessagesDelete?.(event.payload as any);
      })
    );
  }

  // 如果提供了 onUserMessageSaved 处理器，添加用户消息保存事件监听
  if (handlers.onUserMessageSaved) {
    console.log('[listenToStreamEvents] 注册 user_message_saved 事件监听器');
    listeners.push(
      listen('user_message_saved', (event) => {
        console.log('[user_message_saved 事件] 收到事件:', event.payload);
        handlers.onUserMessageSaved?.(event.payload as any);
      })
    );
  }

  const unlisten = await Promise.all(listeners);

  // 返回取消监听的函数
  return () => {
    unlisten.forEach(fn => fn());
  };
}

export async function executeToolCalls(messageId: string, toolCallIds: string[]): Promise<void> {
  await apiCall<void>('message_execute_tool_calls', {
    messageId: messageId,
    toolCallIds: toolCallIds
  });
}

/**
 * 流式执行工具调用
 */
export async function executeToolCallsStream(messageId: string, toolCallIds: string[]): Promise<void> {
  await apiCall<void>('message_execute_tool_calls_stream', {
    messageId: messageId,
    toolCallIds: toolCallIds
  });
}

File: src-tauri/src/commands/message.rs
// 消息相关 IPC 命令

use crate::models::{AppError, Message, MessageRequest, MessageResponse, UUID};
use crate::services::{message::StreamChunk, MessageService};
use serde_json::json;
use tauri::{Emitter, State, Window};

/// 创建流式开始回调
fn create_stream_start_callback(
    window: Window,
    event_name: &'static str,
    extra_data: Option<serde_json::Value>,
) -> impl FnMut(String, String) {
    move |returned_stream_id: String, message_id: String| {
        let mut payload = json!({
            "streamId": returned_stream_id.clone(),
            "messageId": message_id
        });

        // 合并额外的数据
        if let Some(extra) = &extra_data {
            if let Some(obj) = payload.as_object_mut() {
                if let Some(extra_obj) = extra.as_object() {
                    for (key, value) in extra_obj {
                        obj.insert(key.clone(), value.clone());
                    }
                }
            }
        }

        let _ = window.emit(event_name, payload);
        tracing::info!(
            "[{}] Stream started for stream {}",
            event_name,
            returned_stream_id
        );
    }
}

/// 创建流式数据回调
fn create_streaming_callback(window: Window, event_name: &'static str) -> impl FnMut(StreamChunk) {
    move |chunk: StreamChunk| {
        let _ = window.emit(
            event_name,
            json!({
                "streamId": chunk.stream_id,
                "content": chunk.content,
                "reasoning": chunk.reasoning,
                "toolCalls": chunk.tool_calls,
                "chunk": "",
                "index": 0
            }),
        );
    }
}

/// 创建流式结束回调
fn create_stream_end_callback(
    window: Window,
    event_name: &'static str,
) -> impl FnMut(String, MessageResponse) {
    move |stream_id: String, response: MessageResponse| {
        let _ = window.emit(
            event_name,
            json!({
                "streamId": stream_id,
                "messageId": response.message_id,
                "finalContent": response.content,
                "finalReasoning": response.reasoning,
                "chatId": response.chat_id,
                "modelId": response.model_id,
                "providerId": response.provider_id,
                "toolCalls": response.tool_calls
            }),
        );
        tracing::info!("[{}] Stream ended for stream {}", event_name, stream_id);
    }
}

/// 创建流式错误回调
fn create_stream_error_callback(
    window: Window,
    event_name: &'static str,
) -> impl FnMut(String, AppError) {
    move |stream_id: String, error: AppError| {
        let _ = window.emit(
            event_name,
            json!({
                "streamId": stream_id,
                "error": error.message,
                "code": error.code
            }),
        );
        tracing::error!(
            "[{}] Stream error for stream {}: {:?}",
            event_name,
            stream_id,
            error
        );
    }
}

/// 发送聊天消息
#[tauri::command]
pub async fn message_send(
    request: MessageRequest,
    message_service: State<'_, MessageService>,
) -> Result<MessageResponse, AppError> {
    tracing::info!("[message_send] IPC command called");
    match message_service.send_message(request).await {
        Ok(response) => {
            tracing::info!("[message_send] Command completed successfully");
            Ok(response)
        }
        Err(e) => {
            tracing::error!("[message_send] Command failed: {:?}", e);
            Err(e)
        }
    }
}

/// 获取聊天消息
#[tauri::command]
pub async fn message_list(
    chat_id: UUID,
    limit: Option<i32>,
    offset: Option<i32>,
    message_service: State<'_, MessageService>,
) -> Result<Vec<Message>, AppError> {
    tracing::info!("[message_list] IPC command called for chat_id: {}", chat_id);
    match message_service.get_messages(chat_id, limit, offset).await {
        Ok(messages) => {
            tracing::info!(
                "[message_list] Command completed successfully, returned {} messages",
                messages.len()
            );
            tracing::info!("[message_list] Messages: {:?}", messages);
            Ok(messages)
        }
        Err(e) => {
            tracing::error!("[message_list] Command failed: {:?}", e);
            Err(e)
        }
    }
}

/// 获取单条消息
#[tauri::command]
pub async fn message_get(
    message_id: UUID,
    message_service: State<'_, MessageService>,
) -> Result<Message, AppError> {
    tracing::info!(
        "[message_get] IPC command called for message_id: {}",
        message_id
    );
    match message_service.get_message(message_id).await {
        Ok(message) => {
            tracing::info!("[message_get] Command completed successfully");
            Ok(message)
        }
        Err(e) => {
            tracing::error!("[message_get] Command failed: {:?}", e);
            Err(e)
        }
    }
}

/// 更新消息
#[tauri::command]
pub async fn message_update(
    message_id: UUID,
    content: String,
    message_service: State<'_, MessageService>,
) -> Result<Message, AppError> {
    tracing::info!(
        "[message_update] IPC command called for message_id: {}",
        message_id
    );
    match message_service.update_message(message_id, content).await {
        Ok(message) => {
            tracing::info!("[message_update] Command completed successfully");
            Ok(message)
        }
        Err(e) => {
            tracing::error!("[message_update] Command failed: {:?}", e);
            Err(e)
        }
    }
}

/// 删除消息
#[tauri::command]
pub async fn message_delete(
    message_id: UUID,
    message_service: State<'_, MessageService>,
) -> Result<(), AppError> {
    tracing::info!(
        "[message_delete] IPC command called for message_id: {}",
        message_id
    );
    match message_service.delete_message(message_id).await {
        Ok(()) => {
            tracing::info!("[message_delete] Command completed successfully");
            Ok(())
        }
        Err(e) => {
            tracing::error!("[message_delete] Command failed: {:?}", e);
            Err(e)
        }
    }
}

/// 重新生成助手消息
#[tauri::command]
pub async fn message_regenerate(
    message_id: UUID,
    message_service: State<'_, MessageService>,
) -> Result<MessageResponse, AppError> {
    tracing::info!(
        "[message_regenerate] IPC command called for message_id: {}",
        message_id
    );
    match message_service.regenerate_message(message_id).await {
        Ok(response) => {
            tracing::info!("[message_regenerate] Command completed successfully");
            Ok(response)
        }
        Err(e) => {
            tracing::error!("[message_regenerate] Command failed: {:?}", e);
            Err(e)
        }
    }
}

/// 发送流式消息
#[tauri::command]
pub async fn message_send_stream(
    request: MessageRequest,
    window: Window,
    message_service: State<'_, MessageService>,
) -> Result<(), AppError> {
    tracing::info!("[message_send_stream] IPC command called");

    // 克隆必要的数据和窗口引用
    let window_clone = window.clone();
    let request_clone = request.clone();
    let service_clone = message_service.inner().clone();

    // 在后台任务中处理流式响应
    tokio::spawn(async move {
        let start_callback =
            create_stream_start_callback(window_clone.clone(), "message_stream_start", None);

        let streaming_callback =
            create_streaming_callback(window_clone.clone(), "message_stream_chunk");

        let end_callback = create_stream_end_callback(window_clone.clone(), "message_stream_end");

        let error_callback =
            create_stream_error_callback(window_clone.clone(), "message_stream_error");

        // 调用真实的流式API
        service_clone
            .send_message_stream(
                request_clone,
                start_callback,
                streaming_callback,
                end_callback,
                error_callback,
            )
            .await;
    });

    // 立即返回，真实的 stream_id 通过 start 回调事件传递给前端
    Ok(())
}

/// 执行工具调用
#[tauri::command]
pub async fn message_execute_tool_calls(
    message_id: String,
    tool_call_ids: Vec<String>,
    message_service: State<'_, MessageService>,
) -> Result<(), AppError> {
    tracing::info!(
        "[message_execute_tool_calls] IPC command called for message_id: {} with tool call IDs: {:?}",
        message_id,
        tool_call_ids
    );

    message_service
        .execute_tool_calls(
            message_id,
            tool_call_ids,
            |_stream_id, _message_id| {
                // 开始回调
                tracing::info!("[message_execute_tool_calls] Execution started");
            },
            |_chunk| {
                // 工具调用执行的流式输出，暂时不需要特别处理
                // 可以在这里添加日志或进度追踪
            },
            |_stream_id, _response| {
                // 结束回调
                tracing::info!("[message_execute_tool_calls] Execution completed");
            },
            |_stream_id, error| {
                // 错误回调
                tracing::error!("[message_execute_tool_calls] Execution failed: {:?}", error);
            },
        )
        .await;

    tracing::info!("[message_execute_tool_calls] Command completed successfully");
    Ok(())
}

/// 流式执行工具调用
#[tauri::command]
pub async fn message_execute_tool_calls_stream(
    message_id: String,
    tool_call_ids: Vec<String>,
    window: Window,
    message_service: State<'_, MessageService>,
) -> Result<(), AppError> {
    tracing::info!(
        "[message_execute_tool_calls_stream] IPC command called for message_id: {} with tool call IDs: {:?}",
        message_id,
        tool_call_ids
    );

    // 克隆必要的数据和窗口引用
    let window_clone = window.clone();
    let service_clone = message_service.inner().clone();
    let message_id_clone = message_id.clone();
    let tool_call_ids_clone = tool_call_ids.clone();

    // 在后台任务中处理流式响应
    tokio::spawn(async move {
        let start_callback =
            create_stream_start_callback(window_clone.clone(), "tool_execute_stream_start", None);

        let streaming_callback =
            create_streaming_callback(window_clone.clone(), "tool_execute_stream_chunk");

        let end_callback =
            create_stream_end_callback(window_clone.clone(), "tool_execute_stream_end");

        let error_callback =
            create_stream_error_callback(window_clone.clone(), "tool_execute_stream_error");

        // 调用真实的工具执行流式API
        service_clone
            .execute_tool_calls(
                message_id_clone,
                tool_call_ids_clone,
                start_callback,
                streaming_callback,
                end_callback,
                error_callback,
            )
            .await;
    });

    // 立即返回，真实的 stream_id 通过 start 回调事件传递给前端
    Ok(())
}
